semantics/src/main/kotlin/com/strumenta/kolasu/semantics/scope/provider/declarative/DeclarativeScopeProvider.kt

```diff
@@ -0,0 +1,112 @@
+
+/**
+ * Declarative scope provider instances can be used to specify language-specific
+ * scoping rules. Using these, the provider computes scope descriptions
+ * for node reference properties, which can then be used to perform symbol resolution.
+ * Scoping rules can be provided as `DeclarativeScopeProviderRule` instances during instantiation.
+ *
+ * Given a specific language, we suggest to define an object extending this class to specify
+ * language-specific scoping rules:
+ * ```kotlin
+ * object MyScopeProvider: DeclarativeScopeProvider(
+ *     scopeFor(ANode::aReference) {
+ *         name("theSymbolName")
+ *         // local symbols (ScopeDescription API)
+ *         include(somePossiblyNamedNode)
+ *         include("explicitName", someNode)
+ *         // global symbols (ScopeDescription API)
+ *         val symbols: SymbolRepository = ASymbolRepository();
+ *         symbols.allOfType(BNode::class).forEach { include(it) }
+ *         symbols.allOfType(BNode::class).forEach { include( `explicitNameWith${it.someProperty}`, it) }
+ *     }
+ * )
+ * ```
+ **/
+open class DeclarativeScopeProvider(
+    vararg rules: DeclarativeScopeProviderRule<out Node>
+) : ScopeProvider {
+    private val rules: List<DeclarativeScopeProviderRule<out Node>> = rules.sorted()
+
+    override fun <NodeType : Node> scopeFor(
+        node: NodeType,
+        reference: KProperty1<in NodeType, ReferenceByName<out PossiblyNamed>?>
+    ): ScopeDescription {
+        return this.rules.firstOrNull { it.canBeInvokedWith(node::class, reference) }
+            ?.invoke(this, node)
+            ?: throw RuntimeException(
+                "Cannot find scoping rule for reference ${node::class.qualifiedName}::${reference.name}"
+            )
+    }
+}
+
+/**
+ * Represents a scoping rule, i.e. the body of `scopeFor(...)` definitions.
+ **/
+class DeclarativeScopeProviderRule<NodeTy : Node>(
+    private val nodeType: KClass<NodeTy>,
+    private val propertyName: String,
+    private val ignoreCase: Boolean,
+    private val specification: ScopeDescription.(DeclarativeScopeProviderRuleContext<NodeTy>) -> Unit
+) : (ScopeProvider, Node) -> ScopeDescription, Comparable<DeclarativeScopeProviderRule<out Node>> {
+    override fun invoke(
+        scopeProvider: ScopeProvider,
+        node: Node
+    ): ScopeDescription {
+        return ScopeDescription(this.ignoreCase).apply {
+            @Suppress("UNCHECKED_CAST")
+            val context = DeclarativeScopeProviderRuleContext(node as NodeTy, scopeProvider)
+            this.specification(context)
+        }
+    }
+
+    fun canBeInvokedWith(
+        nodeType: KClass<*>,
+        property: KProperty1<*, ReferenceByName<out PossiblyNamed>?>
+    ): Boolean {
+        return this.nodeType.isSuperclassOf(nodeType) && this.propertyName == property.name
+    }
+
+    override fun compareTo(other: DeclarativeScopeProviderRule<out Node>): Int {
```

Comment:
This seems logic we have used elsewhere. Would it make sense to create a utility method called "orderBySpecificity"?

Reply:
There is a similar extension on KClass<*> in `com.strumenta.kolasu.utils` - this works on KType. I could move it there.